src/devices/cddasongloader.cpp

```diff
@@ -30,10 +30,19 @@ CddaSongLoader::CddaSongLoader(const QUrl& url, QObject* parent)
     : QObject(parent), url_(url), cdda_(nullptr), cdio_(nullptr) {
   connect(this, SIGNAL(MusicBrainzDiscIdLoaded(const QString&)),
           SLOT(LoadAudioCDTags(const QString&)));
+
+  cdio_ = cdio_open(url_.path().toLocal8Bit().constData(), DRIVER_DEVICE);
+  Q_ASSERT(cdio_ != nullptr);  // todo: assert for now, but we need to actually
+                               // handle error cases here
 }
 
 CddaSongLoader::~CddaSongLoader() {
-  if (cdio_) cdio_destroy(cdio_);
+  // The LoadSongsFromCdda methods runs concurrently in a thread and we need to
+  // wait for it to terminate. There's no guarantee that it has terminated when
+  // destructor is invoked.
+  loading_future_.waitForFinished();
```

Comment:
Would there be a way to interrupt LoadSongsFromCdda? There are a couple loops that could check for abort flags.

Reply:
Haven't looked into that yet because `LoadSongsFromCdda` shouldn't really take too long to complete in any case, but I can add a flag for it to abort.

Reply:
I have now introduced a `may_load_` flag and have `LoadSongsFromCdda` interrupt its metadata parsing loop if that is no longer set.